### PR TITLE
Allow use of overloaded methods when using LoggerMessage

### LINKED ISSUE #61814: Logging Source Generator fails ungracefully with overloaded methods
### Description

The LoggerMessage source generator fails when there are overloaded logging methods.
### Reproduction Steps

```csharp
public static partial class LoggerMessageGeneration
{
    [LoggerMessage(EventId = 1, EventName = "LogInt", Level = LogLevel.Debug, Message = "Test message {One}")]
    public static partial void Log1(ILogger logger, int one);

    [LoggerMessage(EventId = 2, EventName = "LogBool", Level = LogLevel.Debug, Message = "Test message {Two}")]
    public static partial void Log1(ILogger logger, bool two);
}
```

### Expected behavior

The code compiles successfully. If the LoggerMessage generator explicitly does not support logging method overloads, then a diagnostic should be emitted instead.

Consider the following code, where an overload is specified without an `EventName` - should this compile?

```csharp

public static partial class LoggerMessageGeneration
{
    [LoggerMessage(EventId = 1, Level = LogLevel.Debug, Message = "Test message {One}")]
    public static partial void Log1(ILogger logger, int one);

    [LoggerMessage(EventId = 2, Level = LogLevel.Debug, Message = "Test message {Two}")]
    public static partial void Log1(ILogger logger, bool two);
}
```

The current behaviour of the source generator is to use the method name as the event name, this would emit logs with different event ids that share the same event name. If there are logging providers that rely on event names being unique, this could be problematic.

### Actual behavior

Emits
```
LoggerMessage.g.cs(19,152,19,166): error CS0102: The type 'LoggerMessageGeneration' already contains a definition for '__Log1Callback".
```

### Regression?

No

### Known Workarounds

Do not use overloaded methods when using `[LoggerMessage]`.

### Configuration

```
> dotnet --info

.NET SDK (reflecting any global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100\

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa
```

### Other information

_No response_

### PR DESCRIPTION
**Pull request by JakeYallop**
Fixes #61814

Computes a unique name for `LoggerMethod`s, allowing for overloaded methods to be used with the source generator.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=64573) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=64573) before we can accept your contribution.<br/><br/>:x: JakeYallop [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=64573)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=64573) it.</sub>

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace Microsoft.Extensions.Logging.Generators.Tests.TestClasses
+{
+    internal static partial class OverloadTestExtensions
+    {
+        [LoggerMessage(EventId = 0, Level = LogLevel.Trace, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, int p0);
+
+        [LoggerMessage(EventId = 1, Level = LogLevel.Trace, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, string p0);
+
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs:13]: nit: multiple blank lines

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs [hunk 2]
```diff
@@ -510,6 +510,23 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
...
                         if (lc != null)
                         {
+                            //once we've collected all methods for the given class, check for overloads
+                            //and provide unique names for logger methods
+                            var methods = new Dictionary<string, int>(lc.Methods.Count);
+                            foreach (LoggerMethod lm in lc.Methods)
+                            {
+                                if (methods.ContainsKey(lm.Name))
+                                {
+                                    var currentCount = methods[lm.Name];
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:520]: nit: as per coding style, please use int not var

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.Extensions.Logging.Generators.Tests.TestClasses
+{
+    internal static partial class OverloadTestExtensions
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs:6]: nit: might want to rename filename to `OverloadTestExtensions` to keep the filename and class name in sync

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        [LoggerMessage(EventId = 0, Level = LogLevel.Trace, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, int p0);
+
+        [LoggerMessage(EventId = 1, Level = LogLevel.Trace, Message = "M0{p0}")]
+        public static partial void M0(ILogger logger, string p0);
+
+
+        [LoggerMessage(EventId = 2, EventName = "M1Custom", Level = LogLevel.Trace, Message = "M1{p0}")]
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs:15]: Aren't the first two `M0` enough? are you testing something new in the methods `M1` below? If not, then can perhaps remove.

**Review by maryamariyan — APPROVED** [review]
LGTM. Nice work.

**Comment by maryamariyan** [comment]
This could be merged once you get a chance to apply the minor feedback in this PR.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs (hunk 4 above)
Comment (JakeYallop) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/TestClasses/MethodOverloadTestExtensions.cs:15]: My thinking for this was to verify that custom event names are honoured for each individual overload of a logging method. As the implementation is now, it doesn't do a lot, but if the implementation was to change in future as to how EventNames are processed in the parser, then this test would highlight that issue.

**Comment by JakeYallop** [comment]
@maryamariyan

I've updated the PR based on feedback. The only thing left is the last two LoggerMessage overload tests - I've written a comment with my thinking on that - let me know if you still want me to remove them,

**Comment by maryamariyan** [comment]
@JakeYallop Thanks for the quick turnaround.

I would still be tempted to remove the last two overloads. event name uniqueness is not currently being respected and I was thinking we should fix that first and just leave this PR to focus on overloads for now.

**Comment by maryamariyan** [comment]
Failures in CI seem unrelated and I also found them in https://github.com/dotnet/runtime/pull/64593 and https://github.com/dotnet/runtime/pull/64641

**Comment by eerhardt** [comment]
@maryamariyan - are we sure this needs to be backported to 6.0? It doesn't feel like this limitation is critical to fix in 6, but I guess I'd like your opinion/thoughts.

**Comment by maryamariyan** [comment]
I assume it seems like a blocker to developers wanting to migrate to the logging source generator approach. Looking back at the issue doesn't seem like there'd been a lot of hits with this bug as much as there were for the `@` prefixed issue. So I'd definitely service that one.

**Pull request merged**